File: modules/core/src/main/scala/Stomaton.scala
/*
 * Copyright 2021 Hossein Naderi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edomata.core

import cats.*
import cats.arrow.FunctionK
import cats.data.*
import cats.implicits.*
import cats.kernel.Eq

final case class Stomaton[F[_], -Env, S, R, E, A](
    run: (Env, S) => F[ResponseE[R, E, (S, A)]]
) extends AnyVal {

  /** maps output result */
  def map[B](f: A => B)(using Functor[F]): Stomaton[F, Env, S, R, E, B] =
    Stomaton((env, state) => run(env, state).map(_.map((s, a) => (s, f(a)))))

  /** binds another stomaton to this one. */
  def flatMap[Env2 <: Env, B](
      f: A => Stomaton[F, Env2, S, R, E, B]
  )(using Monad[F]): Stomaton[F, Env2, S, R, E, B] =
    Stomaton((env, state) =>
      run(env, state).flatMap { out =>
        out.result match {
          case Right((newState, a)) =>
            f(a)
              .run(env, newState)
              .map(o =>
                o.copy(notifications = out.notifications ++ o.notifications)
              )
          case Left(errs) => ResponseE(Left(errs), out.notifications).pure
        }
      }
    )

  inline def >>=[Env2 <: Env, B](
      f: A => Stomaton[F, Env2, S, R, E, B]
  )(using Monad[F]): Stomaton[F, Env2, S, R, E, B] = flatMap(f)

  inline def >>[Env2 <: Env, B](
      f: => Stomaton[F, Env2, S, R, E, B]
  )(using Monad[F]): Stomaton[F, Env2, S, R, E, B] = andThen(f)

  inline def andThen[Env2 <: Env, B](
      f: => Stomaton[F, Env2, S, R, E, B]
  )(using Monad[F]): Stomaton[F, Env2, S, R, E, B] = flatMap(_ => f)

  /** creates a new stomaton that translates some input to what this stomaton
    * can understand
    */
  def contramap[Env2](f: Env2 => Env): Stomaton[F, Env2, S, R, E, A] =
    Stomaton((env, state) => run(f(env), state))

  def modify(f: S => S)(using Functor[F]): Stomaton[F, Env, S, R, E, A] =
    Stomaton((env, s0) => run(env, s0).map(_.map((s1, t) => (f(s1), t))))

  def decideS(
      f: S => EitherNec[R, S]
  )(using Monad[F]): Stomaton[F, Env, S, R, E, S] =
    Stomaton((env, s0) =>
      run(env, s0).map(res =>
        res.flatMap((ns, _) => ResponseE(f(ns).map(s => (s, s))))
      )
    )

  def decide[B](
      f: A => EitherNec[R, B]
  )(using Monad[F]): Stomaton[F, Env, S, R, E, B] =
    Stomaton((env, s0) =>
      run(env, s0).map(res =>
        res.flatMap((ns, a) => ResponseE(f(a).map(b => (ns, b))))
      )
    )

  def set(s: S)(using Functor[F]): Stomaton[F, Env, S, R, E, A] =
    Stomaton((env, s0) => run(env, s0).map(_.map((_, a) => (s, a))))

  def handleErrorWith[Env2 <: Env](
      f: NonEmptyChain[R] => Stomaton[F, Env2, S, R, E, A]
  )(using Monad[F]): Stomaton[F, Env2, S, R, E, A] = Stomaton((env, state) =>
    run(env, state).flatMap { out =>
      out.result.fold(
        f(_).run(env, state),
        _ => out.pure[F]
      )
    }
  )
}

object Stomaton extends StomatonInstances, StomatonConstructors

sealed transparent trait StomatonInstances {
  given [F[_]: Monad, Env, S, R, E]
      : MonadError[[t] =>> Stomaton[F, Env, S, R, E, t], NonEmptyChain[R]] =
    new MonadError {

      override def raiseError[A](
          e: NonEmptyChain[R]
      ): Stomaton[F, Env, S, R, E, A] =
        Stomaton.decide(Left(e))

      override def handleErrorWith[A](fa: Stomaton[F, Env, S, R, E, A])(
          f: NonEmptyChain[R] => Stomaton[F, Env, S, R, E, A]
      ): Stomaton[F, Env, S, R, E, A] = fa.handleErrorWith(f)

      type G[T] = Stomaton[F, Env, S, R, E, T]
      type D[T] = DecisionT[F, R, E, T]
      override def pure[A](x: A): G[A] =
        Stomaton.pure(x)

      override def map[A, B](fa: G[A])(f: A => B): G[B] = fa.map(f)

      override def flatMap[A, B](fa: G[A])(f: A => G[B]): G[B] = fa.flatMap(f)

      override def tailRecM[A, B](a: A)(
          f: A => G[Either[A, B]]
      ): G[B] =
        Stomaton((env, s0) =>
          Monad[F].tailRecM(ResponseE.pure((s0, a)): ResponseE[R, E, (S, A)]) {
            res =>
              res.result.fold(
                _ => ???,
                (s, a) =>
                  f(a)
                    .run(env, s)
                    .map(res >> _)
                    .map(o =>
                      o.result match {
                        case Right((newState, Left(a))) =>
                          o.as((newState, a)).asLeft
                        case Right((newState, Right(b))) =>
                          o.as((newState, b)).asRight
                        case Left(errs) =>
                          ResponseE(Left(errs), o.notifications).asRight
                      }
                    )
              )
          }
        )
    }

  given [F[_], Env, S, R, E, T](using
      Eq[(Env, S) => F[ResponseE[R, E, (S, T)]]]
  ): Eq[Stomaton[F, Env, S, R, E, T]] =
    Eq.by(_.run)

  given [F[_], S, R, E, T]
      : Contravariant[[env] =>> Stomaton[F, env, S, R, E, T]] =
    new Contravariant {
      override def contramap[A, B](fa: Stomaton[F, A, S, R, E, T])(
          f: B => A
      ): Stomaton[F, B, S, R, E, T] = fa.contramap(f)
    }

}

sealed transparent trait StomatonConstructors {

  /** constructs an stomaton that outputs a pure value */
  def pure[F[_]: Applicative, Env, S, R, E, T](
      t: T
  ): Stomaton[F, Env, S, R, E, T] =
    Stomaton((_, s) => ResponseE.pure((s, t)).pure)

  /** an stomaton with trivial output */
  def unit[F[_]: Applicative, Env, R, E, N, T]
      : Stomaton[F, Env, R, E, N, Unit] =
    pure(())

  /** constructs an stomaton that evaluates an effect */
  def eval[F[_]: Applicative, Env, S, R, E, T](
      f: F[T]
  ): Stomaton[F, Env, S, R, E, T] =
    Stomaton((_, s) => f.map((s, _).pure))

  /** constructs an stomaton that runs an effect using its input */
  def run[F[_]: Applicative, Env, S, R, E, T](
      f: Env => F[T]
  ): Stomaton[F, Env, S, R, E, T] =
    Stomaton((env, state) => f(env).map((state, _).pure))

  /** constructs an stomaton that outputs the context */
  def context[F[_]: Applicative, Env, S, R, E, T]
      : Stomaton[F, Env, S, R, E, Env] =
    run(_.pure[F])

  /** constructs an stomaton that outputs the current state */
  def state[F[_]: Applicative, Env, S, R, E]: Stomaton[F, Env, S, R, E, S] =
    Stomaton((_, s) => ResponseE.pure((s, s)).pure)

  /** constructs an stomaton that sets the current state */
  def set[F[_]: Applicative, Env, S, R, E](
      s: S
  ): Stomaton[F, Env, S, R, E, Unit] =
    Stomaton((_, _) => ResponseE.pure((s, ())).pure)

  /** constructs an stomaton that modifies current state */
  def modify[F[_]: Applicative, Env, S, R, E](
      f: S => S
  ): Stomaton[F, Env, S, R, E, S] =
    Stomaton((_, s) =>
      val ns = f(s)
      ResponseE.pure((ns, ns)).pure
    )

  def decideS[F[_]: Applicative, Env, S, R, E](
      f: S => EitherNec[R, S]
  ): Stomaton[F, Env, S, R, E, S] =
    Stomaton((env, s0) => ResponseE(f(s0).map(s => (s, s))).pure)

  def decide[F[_]: Applicative, Env, S, R, E, T](
      f: => EitherNec[R, T]
  ): Stomaton[F, Env, S, R, E, T] =
    Stomaton((_, s) => ResponseE(f.map(t => (s, t))).pure)

  /** constructs an stomaton that decides to modify state based on current state
    */
  def modifyS[F[_]: Applicative, Env, S, R, E](
      f: S => EitherNec[R, S]
  ): Stomaton[F, Env, S, R, E, S] =
    Stomaton((_, s) => ResponseE(f(s).map(ns => (ns, ns))).pure)

  /** constructs an stomaton that rejects with given rejections */
  def reject[F[_]: Applicative, Env, S, R, E, T](
      r: R,
      rs: R*
  ): Stomaton[F, Env, S, R, E, T] = decide(NonEmptyChain(r, rs: _*).asLeft)

  def publish[F[_]: Applicative, Env, S, R, E](
      ns: E*
  ): Stomaton[F, Env, S, R, E, Unit] =
    Stomaton((_, s) => ResponseE(Right((s, ())), Chain(ns: _*)).pure)

  def validate[F[_]: Applicative, Env, S, R, E, T](
      v: ValidatedNec[R, T]
  ): Stomaton[F, Env, S, R, E, T] = decide(v.toEither)

  /** Constructs a program from an optional value, that outputs value if exists
    * or rejects otherwise
    *
    * You can also use .toDecision syntax for more convenience
    */
  def fromOption[F[_]: Applicative, Env, S, R, E, T](
      opt: Option[T],
      orElse: R,
      other: R*
  ): Stomaton[F, Env, S, R, E, T] =
    opt.fold(reject(orElse, other: _*))(pure(_))

  /** Constructs a program that either outputs a value or rejects
    */
  def fromEither[F[_]: Applicative, Env, S, R, E, T](
      eit: Either[R, T]
  ): Stomaton[F, Env, S, R, E, T] = eit.fold(reject(_), pure(_))

  /** Constructs a program that either outputs a value or rejects
    *
    * You can also use .toDecision syntax for more convenience.
    */
  def fromEitherNec[F[_]: Applicative, Env, S, R, E, T](
      eit: EitherNec[R, T]
  ): Stomaton[F, Env, S, R, E, T] = decide(eit)

}

File: rust/crates/edomata-core/src/stomaton.rs
//! [`Stomaton`]: the state automaton used for CQRS (no event sourcing).

use std::fmt;
use std::future::Future;
use std::sync::Arc;

use crate::{BoxFuture, NonEmpty, ResponseE};

/// The reusable program behind a [`Stomaton`].
type RunFn<Env, S, R, N, A> =
    dyn Fn(Env, S) -> BoxFuture<'static, ResponseE<R, N, (S, A)>> + Send + Sync;

/// A state automaton: a reusable, asynchronous program that reads an input
/// (`Env`) and the current state, produces a new state and an output, and
/// optionally publishes notifications. Unlike [`Edomaton`](crate::Edomaton)
/// it has no domain events: the state is stored directly.
///
/// # Type parameters
///
/// * `Env` – input type, usually a [`CommandMessage`](crate::CommandMessage)
/// * `S` – state type
/// * `R` – rejection type
/// * `N` – notification type
/// * `A` – output type
///
/// ```
/// use edomata_core::{Stomaton, nonempty};
///
/// let inc: Stomaton<i32, i32, &str, &str, i32> = Stomaton::context()
///     .and_then(|by| Stomaton::modify(move |s| s + by))
///     .publish(["incremented"]);
///
/// let out = futures::executor::block_on(inc.run(5, 10));
/// assert_eq!(out.result, Ok((15, 15)));
/// assert_eq!(out.notifications, vec!["incremented"]);
/// ```
pub struct Stomaton<Env, S, R, N, A> {
    run: Arc<RunFn<Env, S, R, N, A>>,
}

impl<Env, S, R, N, A> Clone for Stomaton<Env, S, R, N, A> {
    fn clone(&self) -> Self {
        Self {
            run: Arc::clone(&self.run),
        }
    }
}

impl<Env, S, R, N, A> fmt::Debug for Stomaton<Env, S, R, N, A> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("Stomaton(<program>)")
    }
}

impl<Env, S, R, N, A> Stomaton<Env, S, R, N, A>
where
    Env: Send + 'static,
    S: Send + 'static,
    R: Send + 'static,
    N: Send + 'static,
    A: Send + 'static,
{
    // ---------------------------------------------------------------------
    // Constructors
    // ---------------------------------------------------------------------

    /// Constructs a stomaton from an asynchronous function of its input and
    /// state.
    pub fn new<F, Fut>(f: F) -> Self
    where
        F: Fn(Env, S) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = ResponseE<R, N, (S, A)>> + Send + 'static,
    {
        Self {
            run: Arc::new(move |env, s| Box::pin(f(env, s))),
        }
    }

    /// Constructs a stomaton from a pure function of its input and state.
    pub fn from_fn<F>(f: F) -> Self
    where
        F: Fn(Env, S) -> ResponseE<R, N, (S, A)> + Send + Sync + 'static,
    {
        Self::new(move |env, s| std::future::ready(f(env, s)))
    }

    /// Constructs a stomaton that outputs a pure value.
    pub fn pure(a: A) -> Self
    where
        A: Clone + Sync,
    {
        Self::from_fn(move |_, s| ResponseE::pure((s, a.clone())))
    }

    /// Constructs a stomaton that evaluates an effect, whose output becomes
    /// the program output.
    pub fn eval<F, Fut>(f: F) -> Self
    where
        F: Fn() -> Fut + Send + Sync + 'static,
        Fut: Future<Output = A> + Send + 'static,
    {
        Self::new(move |_, s| {
            let fut = f();
            async move { ResponseE::pure((s, fut.await)) }
        })
    }

    /// Constructs a stomaton that runs an effect using its input.
    pub fn run_with<F, Fut>(f: F) -> Self
    where
        F: Fn(Env) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = A> + Send + 'static,
    {
        Self::new(move |env, s| {
            let fut = f(env);
            async move { ResponseE::pure((s, fut.await)) }
        })
    }

    /// Constructs a stomaton that outputs a pure function of its input.
    pub fn map_input<F>(f: F) -> Self
    where
        F: Fn(Env) -> A + Send + Sync + 'static,
    {
        Self::from_fn(move |env, s| ResponseE::pure((s, f(env))))
    }

    /// Constructs a stomaton that outputs a value or rejects, without
    /// touching the state.
    pub fn decide(result: Result<A, NonEmpty<R>>) -> Self
    where
        Result<A, NonEmpty<R>>: Clone + Sync,
    {
        Self::from_fn(move |_, s| ResponseE::lift(result.clone().map(|a| (s, a))))
    }

    /// Constructs a stomaton that decides with a function of the current
    /// state, without touching the state.
    pub fn decide_with<F>(f: F) -> Self
    where
        F: Fn(&S) -> Result<A, NonEmpty<R>> + Send + Sync + 'static,
    {
        Self::from_fn(move |_, s| {
            let out = f(&s);
            ResponseE::lift(out.map(|a| (s, a)))
        })
    }

    /// Constructs a stomaton that rejects with the given reasons.
    pub fn reject(reasons: impl Into<NonEmpty<R>>) -> Self
    where
        R: Clone + Sync,
    {
        let reasons = reasons.into();
        Self::from_fn(move |_, _| ResponseE::reject(reasons.clone()))
    }

    /// Constructs a stomaton from a validation.
    pub fn validate(validation: Result<A, NonEmpty<R>>) -> Self
    where
        Result<A, NonEmpty<R>>: Clone + Sync,
    {
        Self::decide(validation)
    }

    /// Constructs a stomaton from an optional value, that outputs the value
    /// if it exists or rejects otherwise.
    pub fn from_option(opt: Option<A>, or_else: impl Into<NonEmpty<R>>) -> Self
    where
        Result<A, NonEmpty<R>>: Clone + Sync,
    {
        Self::decide(opt.ok_or_else(|| or_else.into()))
    }

    /// Constructs a stomaton that either outputs a value or rejects with a
    /// single reason.
    pub fn from_result(result: Result<A, R>) -> Self
    where
        Result<A, NonEmpty<R>>: Clone + Sync,
    {
        Self::decide(result.map_err(NonEmpty::new))
    }

    /// Constructs a stomaton that either outputs a value or rejects with
    /// several reasons.
    pub fn from_result_nec(result: Result<A, NonEmpty<R>>) -> Self
    where
        Result<A, NonEmpty<R>>: Clone + Sync,
    {
        Self::decide(result)
    }

    // ---------------------------------------------------------------------
    // Running
    // ---------------------------------------------------------------------

    /// Runs this stomaton with the given input and state.
    pub fn run(&self, env: Env, state: S) -> BoxFuture<'static, ResponseE<R, N, (S, A)>> {
        (self.run)(env, state)
    }

    // ---------------------------------------------------------------------
    // Combinators
    // ---------------------------------------------------------------------

    /// Transforms the underlying response.
    pub fn transform<R2, N2, B, F>(self, f: F) -> Stomaton<Env, S, R2, N2, B>
    where
        R2: Send + 'static,
        N2: Send + 'static,
        B: Send + 'static,
        F: Fn(ResponseE<R, N, (S, A)>) -> ResponseE<R2, N2, (S, B)> + Send + Sync + 'static,
    {
        let f = Arc::new(f);
        Stomaton::new(move |env, s| {
            let fut = self.run(env, s);
            let f = Arc::clone(&f);
            async move { f(fut.await) }
        })
    }

    /// Maps the output.
    pub fn map<B, F>(self, f: F) -> Stomaton<Env, S, R, N, B>
    where
        B: Send + 'static,
        F: Fn(A) -> B + Send + Sync + 'static,
    {
        self.transform(move |r| r.map(|(s, a)| (s, f(a))))
    }

    /// Replaces the output.
    pub fn replace<B>(self, b: B) -> Stomaton<Env, S, R, N, B>
    where
        B: Clone + Send + Sync + 'static,
    {
        self.map(move |_| b.clone())
    }

    /// Ignores the output.
    pub fn void(self) -> Stomaton<Env, S, R, N, ()> {
        self.map(|_| ())
    }

    /// Creates a new stomaton that translates some input to what this one
    /// can understand.
    pub fn contramap<Env2, F>(self, f: F) -> Stomaton<Env2, S, R, N, A>
    where
        Env2: Send + 'static,
        F: Fn(Env2) -> Env + Send + Sync + 'static,
    {
        Stomaton::new(move |env2, s| self.run(f(env2), s))
    }

    /// Binds another stomaton to this one.
    ///
    /// The next stomaton runs with the state produced by this one.
    /// Notifications of both are kept in order; a rejection of this stomaton
    /// short-circuits and keeps its notifications.
    pub fn and_then<B, F>(self, f: F) -> Stomaton<Env, S, R, N, B>
    where
        Env: Clone,
        B: Send + 'static,
        F: Fn(A) -> Stomaton<Env, S, R, N, B> + Send + Sync + 'static,
    {
        let f = Arc::new(f);
        Stomaton::new(move |env: Env, s| {
            let first = self.run(env.clone(), s);
            let f = Arc::clone(&f);
            async move {
                let out = first.await;
                match out.result {
                    Ok((new_state, a)) => {
                        let o = f(a).run(env, new_state).await;
                        let mut notifications = out.notifications;
                        notifications.extend(o.notifications);
                        ResponseE::new(o.result, notifications)
                    }
                    Err(errs) => ResponseE::new(Err(errs), out.notifications),
                }
            }
        })
    }

    /// Alias of [`Stomaton::and_then`].
    pub fn flat_map<B, F>(self, f: F) -> Stomaton<Env, S, R, N, B>
    where
        Env: Clone,
        B: Send + 'static,
        F: Fn(A) -> Stomaton<Env, S, R, N, B> + Send + Sync + 'static,
    {
        self.and_then(f)
    }

    /// Sequences another stomaton after this one, ignoring this output.
    pub fn then<B>(self, next: Stomaton<Env, S, R, N, B>) -> Stomaton<Env, S, R, N, B>
    where
        Env: Clone,
        B: Send + 'static,
    {
        self.and_then(move |_| next.clone())
    }

    /// Modifies the resulting state.
    pub fn modify_state<F>(self, f: F) -> Self
    where
        F: Fn(S) -> S + Send + Sync + 'static,
    {
        self.transform(move |r| r.map(|(s, a)| (f(s), a)))
    }

    /// Decides on a new state from the resulting state; the output becomes
    /// the new state.
    pub fn decide_state<F>(self, f: F) -> Stomaton<Env, S, R, N, S>
    where
        S: Clone,
        F: Fn(S) -> Result<S, NonEmpty<R>> + Send + Sync + 'static,
    {
        self.transform(move |r| {
            r.and_then(|(s, _)| ResponseE::lift(f(s).map(|ns| (ns.clone(), ns))))
        })
    }

    /// Decides on the output.
    pub fn decide_output<B, F>(self, f: F) -> Stomaton<Env, S, R, N, B>
    where
        B: Send + 'static,
        F: Fn(A) -> Result<B, NonEmpty<R>> + Send + Sync + 'static,
    {
        self.transform(move |r| r.and_then(|(s, a)| ResponseE::lift(f(a).map(|b| (s, b)))))
    }

    /// Sets the resulting state.
    pub fn set_state(self, state: S) -> Self
    where
        S: Clone + Sync,
    {
        self.transform(move |r| r.map(|(_, a)| (state.clone(), a)))
    }

    /// Recovers from a rejection, running `f` with the original state.
    pub fn handle_error_with<F>(self, f: F) -> Self
    where
        Env: Clone,
        S: Clone,
        F: Fn(NonEmpty<R>) -> Stomaton<Env, S, R, N, A> + Send + Sync + 'static,
    {
        let f = Arc::new(f);
        Stomaton::new(move |env: Env, s: S| {
            let first = self.run(env.clone(), s.clone());
            let f = Arc::clone(&f);
            async move {
                let out = first.await;
                match out.result {
                    Err(errs) => f(errs).run(env, s).await,
                    Ok(v) => ResponseE::new(Ok(v), out.notifications),
                }
            }
        })
    }

    /// Adds notifications regardless of the decision state.
    pub fn publish<I>(self, ns: I) -> Self
    where
        I: IntoIterator<Item = N> + Clone + Send + Sync + 'static,
    {
        self.transform(move |r| r.publish(ns.clone()))
    }

    /// If this stomaton is rejected, uses `f` to decide what to publish.
    pub fn publish_on_rejection_with<F, I>(self, f: F) -> Self
    where
        F: Fn(&NonEmpty<R>) -> I + Send + Sync + 'static,
        I: IntoIterator<Item = N>,
    {
        self.transform(move |r| r.publish_on_rejection_with(&f))
    }

    /// Publishes the given notifications if this stomaton is rejected.
    pub fn publish_on_rejection<I>(self, ns: I) -> Self
    where
        I: IntoIterator<Item = N> + Clone + Send + Sync + 'static,
    {
        self.transform(move |r| r.publish_on_rejection(ns.clone()))
    }
}

impl<Env, S, R, N> Stomaton<Env, S, R, N, ()>
where
    Env: Send + 'static,
    S: Send + 'static,
    R: Send + 'static,
    N: Send + 'static,
{
    /// A stomaton with a trivial output.
    pub fn unit() -> Self {
        Self::from_fn(|_, s| ResponseE::pure((s, ())))
    }

    /// Constructs a stomaton that sets the current state.
    pub fn set(state: S) -> Self
    where
        S: Clone + Sync,
    {
        Self::from_fn(move |_, _| ResponseE::pure((state.clone(), ())))
    }

    /// Constructs a stomaton that publishes the given notifications.
    pub fn publish_only<I>(ns: I) -> Self
    where
        I: IntoIterator<Item = N> + Clone + Send + Sync + 'static,
    {
        Self::from_fn(move |_, s| ResponseE::new(Ok((s, ())), ns.clone()))
    }
}

impl<Env, S, R, N> Stomaton<Env, S, R, N, S>
where
    Env: Send + 'static,
    S: Clone + Send + 'static,
    R: Send + 'static,
    N: Send + 'static,
{
    /// Constructs a stomaton that outputs the current state.
    pub fn state() -> Self {
        Self::from_fn(|_, s: S| ResponseE::pure((s.clone(), s)))
    }

    /// Constructs a stomaton that modifies the current state and outputs the
    /// new state.
    pub fn modify<F>(f: F) -> Self
    where
        F: Fn(S) -> S + Send + Sync + 'static,
    {
        Self::from_fn(move |_, s| {
            let ns = f(s);
            ResponseE::pure((ns.clone(), ns))
        })
    }

    /// Constructs a stomaton that decides to modify the state based on the
    /// current state and outputs the new state.
    pub fn modify_s<F>(f: F) -> Self
    where
        F: Fn(S) -> Result<S, NonEmpty<R>> + Send + Sync + 'static,
    {
        Self::from_fn(move |_, s| ResponseE::lift(f(s).map(|ns| (ns.clone(), ns))))
    }

    /// Alias of [`Stomaton::modify_s`].
    pub fn decide_s<F>(f: F) -> Self
    where
        F: Fn(S) -> Result<S, NonEmpty<R>> + Send + Sync + 'static,
    {
        Self::modify_s(f)
    }
}

impl<Env, S, R, N> Stomaton<Env, S, R, N, Env>
where
    Env: Send + 'static,
    S: Send + 'static,
    R: Send + 'static,
    N: Send + 'static,
{
    /// Constructs a stomaton that outputs its input.
    pub fn context() -> Self {
        Self::from_fn(|env, s| ResponseE::pure((s, env)))
    }
}
